compiler/src/iree/compiler/Codegen/Common/TypePropagationPass.cpp

```diff
@@ -429,6 +429,91 @@ struct IREELinalgExtSortTypePropagation
   }
 };
 
+/// Pattern to legalize `linalg.reduce` operations.
+struct ReduceOpTypePropagation
+    : public TypePropagationPattern<linalg::ReduceOp> {
+  using TypePropagationPattern<linalg::ReduceOp>::TypePropagationPattern;
+
+  LogicalResult
+  matchAndRewrite(linalg::ReduceOp reduceOp, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+    bool needsConversion = false;
+    for (auto &operand : reduceOp->getOpOperands()) {
+      Type operandType = operand.get().getType();
+      Type legalizedType = this->getTypeConverter()->convertType(operandType);
+      if (operandType != legalizedType) {
+        needsConversion = true;
+        break;
+      }
+    }
+
+    if (!needsConversion) {
+      return rewriter.notifyMatchFailure(
+          reduceOp, "all types already legal within conversion pattern");
+    }
+
+    SmallVector<Type> resultTypes;
+    for (auto init : reduceOp.getInits()) {
+      Type legalizedType =
+          this->getTypeConverter()->convertType(init.getType());
+      resultTypes.push_back(legalizedType);
+    }
+
+    auto modifiedOp = cast<linalg::ReduceOp>(mlir::cloneWithoutRegions(
+        rewriter, reduceOp.getOperation(), resultTypes, adaptor.getOperands()));
+
+    rewriter.inlineRegionBefore(reduceOp.getCombiner(),
+                                modifiedOp.getCombiner(),
+                                modifiedOp.getCombiner().begin());
+    Region &modifiedOpRegion = modifiedOp.getCombiner();
+
+    TypeConverter::SignatureConversion signatureConverter(
+        modifiedOpRegion.getNumArguments());
+    for (auto [index, arg] : llvm::enumerate(modifiedOpRegion.getArguments())) {
+      Type argType = arg.getType();
+      std::optional<Type> legalizedArgType =
+          legalizeStorageElementType(argType);
+      if (!legalizedArgType) {
+        return reduceOp.emitOpError("failed to get legalized type for arg ")
+               << index;
+      }
+      signatureConverter.addInputs(index, legalizedArgType.value());
+    }
+    rewriter.applySignatureConversion(&modifiedOpRegion.front(),
+                                      signatureConverter, getTypeConverter());
+
+    OpBuilder::InsertionGuard g(rewriter);
+    Block *entryBlock = &modifiedOpRegion.front();
+    Operation *yieldOp = entryBlock->getTerminator();
+    rewriter.setInsertionPoint(yieldOp);
+
+    SmallVector<Value> convertedYieldValues;
+    bool needsUpdate = false;
+    for (Value yieldValue : yieldOp->getOperands()) {
+      Type yieldType = yieldValue.getType();
+      std::optional<Type> legalizedYieldType =
+          legalizeStorageElementType(yieldType);
+
+      Value valueToYield = yieldValue; // Default to original value
```

Comment:
```suggestion
      Value valueToYield = yieldValue; // Default to original value.
```
https://llvm.org/docs/CodingStandards.html#commenting